File: apps/lazuar-pay/src/Lazuar.Pay/Gateways/BuyerEmail.cs
namespace Lazuar.Pay.Gateways;

public static class BuyerEmail
{
    public const string Placeholder = "customer@example.com";

    public static bool IsUsable(string? email) =>
        !string.IsNullOrWhiteSpace(email)
        && !string.Equals(email.Trim(), Placeholder, StringComparison.OrdinalIgnoreCase);

    public static string NameFrom(string? email, string? name)
    {
        if (!string.IsNullOrWhiteSpace(name))
        {
            return name.Trim();
        }

        if (string.IsNullOrWhiteSpace(email))
        {
            return "Customer";
        }

        var at = email.IndexOf('@');
        return at > 0 ? email[..at] : "Customer";
    }
}

File: apps/lazuar-api/src/publicpay/buyer_email.rs
//! Port of `PublicPay/BuyerEmail.cs`.

pub const PLACEHOLDER: &str = "customer@example.com";

pub fn is_usable(email: Option<&str>) -> bool {
    match email {
        Some(e) => {
            !e.trim().is_empty() && !e.trim().eq_ignore_ascii_case(PLACEHOLDER)
        }
        None => false,
    }
}

/// `BuyerEmail.NameFrom` — display name preference: payer name, else email local part.
pub fn name_from(email: Option<&str>, name: Option<&str>) -> String {
    if let Some(n) = name.map(str::trim).filter(|s| !s.is_empty()) {
        return n.to_string();
    }
    match email.map(str::trim).filter(|s| !s.is_empty()) {
        Some(e) => e.split('@').next().unwrap_or(e).to_string(),
        None => "Customer".to_string(),
    }
}

/// `NormalizeSlotKey` — trimmed, 8..=128 chars, else absent.
pub fn normalize_slot_key(raw: Option<&str>) -> Option<String> {
    let slot = raw?.trim();
    if slot.is_empty() || !(8..=128).contains(&slot.len()) {
        return None;
    }
    Some(slot.to_string())
}
